js/player_thread.js

```diff
@@ -0,0 +1,475 @@
+/* eslint-disable no-underscore-dangle */
+/* global navigator MediaMetadata */
+/* global Howl Howler */
+
+{
+  const sendEvent = (event) => {
+    (chrome || browser).runtime.sendMessage(event);
+  };
+
+  /**
+   * Player class containing the state of our playlist and where we are in it.
+   * Includes all methods for playing, skipping, updating the display, etc.
+   * @param {Array} playlist Array of objects with playlist song details ({title, file, howl}).
+   */
+  class Player {
+    constructor() {
+      this.playlist = [];
+      this.index = -1;
+      this._loop_mode = 0;
+      this._media_uri_list = {};
+      this.playedFrom = 0;
+    }
+
+    setRefreshRate(rate = 10) {
+      clearInterval(this.refreshTimer);
+      this.refreshTimer = setInterval(() => {
+        if (this.playing) {
+          this.sendFrameUpdate();
+        }
+      }, 1000 / rate);
+    }
+
+    get currentAudio() {
+      return this.playlist[this.index];
+    }
+
+    get currentHowl() {
+      return this.currentAudio && this.currentAudio.howl;
+    }
+
+    get playing() {
+      return this.currentHowl ? this.currentHowl.playing() : false;
+    }
+
+    static get muted() {
+      return !!Howler._muted;
+    }
+
+    insertAudio(audio, idx) {
```

Comment:
avoid add same song to playlist multiple times.

      if (this.playlist.findIndex(a=> audio.id === a.id) > -1){
        return;
      }

I think it's better to use track.id for play parameter, because current method reply on position of song in playlist, but sometimes playlist will change durning it take effect.